google-cloud-firestore/src/main/java/com/google/cloud/firestore/AggregateQuery.java

```diff
@@ -75,58 +76,134 @@ public ApiFuture<AggregateQuerySnapshot> get() {
+            /* transactionId= */ null,
+            /* readTime= */ null,
+            /* startTimeNanos= */ query.rpcContext.getClock().nanoTime(),
+            /* explainOptions= */ options);
+    runQuery(responseDeliverer);
+    return responseDeliverer.getExplainResultsFuture();
+  }
+
   @Nonnull
   ApiFuture<AggregateQuerySnapshot> get(
       @Nullable final ByteString transactionId, @Nullable com.google.protobuf.Timestamp readTime) {
     AggregateQueryResponseDeliverer responseDeliverer =
         new AggregateQueryResponseDeliverer(
-            transactionId, readTime, /* startTimeNanos= */ query.rpcContext.getClock().nanoTime());
+            transactionId,
+            readTime,
+            /* startTimeNanos= */ query.rpcContext.getClock().nanoTime(),
+            /* explainOptions= */ null);
     runQuery(responseDeliverer);
-    return responseDeliverer.getFuture();
+    return responseDeliverer.getAggregateQuerySnapshotFuture();
   }
 
   private void runQuery(AggregateQueryResponseDeliverer responseDeliverer) {
     RunAggregationQueryRequest request =
-        toProto(responseDeliverer.transactionId, responseDeliverer.readTime);
+        toProto(
+            responseDeliverer.transactionId,
+            responseDeliverer.readTime,
+            responseDeliverer.explainOptions);
     AggregateQueryResponseObserver responseObserver =
         new AggregateQueryResponseObserver(responseDeliverer);
     ServerStreamingCallable<RunAggregationQueryRequest, RunAggregationQueryResponse> callable =
         query.rpcContext.getClient().runAggregationQueryCallable();
     query.rpcContext.streamRequest(request, responseObserver, callable);
   }
 
+  @Nonnull
+  private Map<String, Value> convertServerAggregateFieldsMapToClientAggregateFieldsMap(
+      @Nonnull Map<String, Value> data) {
+    ImmutableMap.Builder<String, Value> builder = ImmutableMap.builder();
+    data.forEach((serverAlias, value) -> builder.put(aliasMap.get(serverAlias), value));
+    return builder.build();
+  }
+
   private final class AggregateQueryResponseDeliverer {
 
     @Nullable private final ByteString transactionId;
     @Nullable private final com.google.protobuf.Timestamp readTime;
     private final long startTimeNanos;
-    private final SettableApiFuture<AggregateQuerySnapshot> future = SettableApiFuture.create();
-    private final AtomicBoolean isFutureCompleted = new AtomicBoolean(false);
+    @Nullable private final ExplainOptions explainOptions;
+    private final SettableApiFuture<AggregateQuerySnapshot> aggregateQuerySnapshotFuture =
+        SettableApiFuture.create();
+    private final SettableApiFuture<ExplainResults<AggregateQuerySnapshot>> explainFuture =
+        SettableApiFuture.create();
+    private final AtomicBoolean isAggregateQuerySnapshotFutureCompleted = new AtomicBoolean(false);
+    private final AtomicBoolean isExplainFutureCompleted = new AtomicBoolean(false);
 
     AggregateQueryResponseDeliverer(
```

Comment:
There is a clear line dividing all paths of logic in this class: `if (explainOptions != null)`.

Should we maybe just have separate classes for these separate purposes. I am not sure whether this should be done with inheritance, interfaces, abstract classes, restructuring logic to AggregateQueryResponseObserver.

Current implementation seems awkward though, with some fields only be used in one path, and other fields only being used in other path.

Reply:
I agree.